runtime/lua/vim/lsp/document_highlight.lua

```diff
@@ -0,0 +1,323 @@
+      local client_hilights = bufstate.client_hilights or {}
+
+      if not client_hilights[client.id] then
+        client_hilights[client.id] = {}
+      end
+
+      bufstate.client_hilights = client_hilights
+      bufstates[bufnr] = bufstate
+    end
+  end
+end
+api.nvim_create_autocmd('LspAttach', {
+  group = augroup,
+  callback = function(args)
+    ---@type integer
+    local client_id = args.data.client_id
+    if
+      not assert(vim.lsp.get_client_by_id(client_id)).supports_method(
+        ms.textDocument_documentHighlight
+      )
+    then
+      return
+    end
+
+    ---@type integer
+    local bufnr = args.buf
+    local bufstate = bufstates[bufnr] or {}
+    bufstates[bufnr] = bufstate
+
+    local client_hilights = bufstate.client_hilights or {}
+    client_hilights[client_id] = {}
+    bufstate.client_hilights = client_hilights
+
+    api.nvim_buf_attach(bufnr, false, {
+      on_detach = function()
+        bufstates[bufnr] = nil
+      end,
+    })
+  end,
+})
+api.nvim_create_autocmd('LspDetach', {
+  group = augroup,
+  callback = function(args)
+    -- bufnr may not exists if the buffer is already unloaded
+    ---@type integer?
+    local bufnr = args.bufnr
+    if not bufnr then
+      return
+    end
+    local bufstate = assert(bufstates[bufnr])
+
+    ---@type integer
+    local client_id = args.data.client_id
+    bufstate.client_hilights[client_id] = nil
+    if vim.tbl_isempty(bufstate.client_hilights) then
+      bufstates[bufnr] = nil
+    end
+  end,
+})
+
+---|lsp-handler| for the method `textDocument/documentHighlight`
+---Store hints for a specific buffer and client
+---@param result? lsp.DocumentHighlight[]
+---@param ctx lsp.HandlerContext
+function M.on_document_highlight(err, result, ctx)
+  if err then
+    log.error('document highlight', err)
+  end
+
+  local bufnr = assert(ctx.bufnr)
+  local bufstate = assert(bufstates[bufnr])
+
+  local client_id = assert(ctx.client_id)
+  local client_hilights = bufstate.client_hilights
+
+  ---@type table<integer, lsp.DocumentHighlight[]?>
+  local row_highlights = {}
+  for _, highlight in pairs(result or {}) do
+    for row = highlight.range['start'].line, highlight.range['end'].line do
+      local highlights = row_highlights[row] or {}
+      highlights[#highlights + 1] = highlight
+      row_highlights[row] = highlights
+    end
+  end
+  client_hilights[client_id] = row_highlights
+  api.nvim__redraw({ buf = bufnr, valid = true })
+end
+
+---@param bufnr integer
+local function refresh(bufnr)
+  local bufstate = assert(bufstates[bufnr])
+  local enabled = bufstate.enabled
+  if enabled == nil then
+    enabled = globalstate.enabled
+  end
+
+  if not enabled then
+    api.nvim_buf_clear_namespace(bufnr, namespace, 0, -1)
+    api.nvim__redraw({ buf = bufnr, valid = true })
+    return
+  end
+
+  local params = util.make_position_params()
+  vim.lsp.buf_request(bufnr, ms.textDocument_documentHighlight, params)
```

Comment:
Params would need to be created per client with `client.offset_encoding`

Reply:
Thank you for your reminder. I haven't completed the work related to offset_encoding yet. I will look into this.